test/NuGet.Core.Tests/NuGet.Packaging.Test/NoRefOrLibFolderInPackageRuleTests.cs

```diff
@@ -0,0 +1,324 @@
+        {
+            var files = new[]
+            {
+                "build/random_tfm/test.dll",
+                "content/random_tfm/test.dll"
+            };
+
+            using (var testContext = TestPackageFile.Create(files, _nuspecContent))
+            {
+                // Act
+                var rule = new NoRefOrLibFolderInPackageRule(AnalysisResources.NoRefOrLibFolderInPackage);
+                var issues = rule.Validate(testContext.PackageArchiveReader).ToList();
+
+                // Assert
+                Assert.True(issues.Any(p => p.Code == NuGetLogCode.NU5127));
+            }
+        }
+
+        [Fact]
+        public void Validate_PackageWithBuildFolderAndFile_ShouldWarn()
+        {
+            var files = new[]
+            {
+                "build/random_tfm/test.dll",
+                "test.dll"
+            };
+
+            using (var testContext = TestPackageFile.Create(files, _nuspecContent))
+            {
+                // Act
+                var rule = new NoRefOrLibFolderInPackageRule(AnalysisResources.NoRefOrLibFolderInPackage);
+                var issues = rule.Validate(testContext.PackageArchiveReader).ToList();
+
+                // Assert
+                Assert.True(issues.Any(p => p.Code == NuGetLogCode.NU5127));
+            }
+        }
+
+        [Fact]
+        public void Validate_PackageWithLibFolderAndFile_ShouldNotWarn()
+        {
+            var files = new[]
+            {
+                "lib/random_tfm/test.dll",
+                "test.dll"
+            };
+
+            using (var testContext = TestPackageFile.Create(files, _nuspecContent))
+            {
+                // Act
+                var rule = new NoRefOrLibFolderInPackageRule(AnalysisResources.NoRefOrLibFolderInPackage);
+                var issues = rule.Validate(testContext.PackageArchiveReader).ToList();
+
+                // Assert
+                Assert.False(issues.Any(p => p.Code == NuGetLogCode.NU5127));
+            }
+        }
+
+        [Fact]
+        public void Validate_PackageWithRefFolderAndFile_ShouldNotWarn()
+        {
+            var files = new[]
+            {
+                "ref/random_tfm/test.dll",
+                "test.dll"
+            };
+
+            using (var testContext = TestPackageFile.Create(files, _nuspecContent))
+            {
+                // Act
+                var rule = new NoRefOrLibFolderInPackageRule(AnalysisResources.NoRefOrLibFolderInPackage);
+                var issues = rule.Validate(testContext.PackageArchiveReader).ToList();
+
+                // Assert
+                Assert.False(issues.Any(p => p.Code == NuGetLogCode.NU5127));
+            }
+        }
+    }
+
+    internal class TestPackageFile : IDisposable
+    {
+        private TestPackageFile(PackageArchiveReader reader)
+        {
+            PackageArchiveReader = reader;
+        }
+
+        public static TestPackageFile Create(IEnumerable<string> files, string nuspecContent)
```

Comment:
You can also make a bunch of these tests cheaper by refactoring the validation method to take only the file list as input and test that for most of the tests, and then do only 1 or 2 full blown PackageArchiveReader tests. 

Note that you can keep that refactored method internal.

Reply:
If I do that, could I still dispose of the `TestPackageFile`?

Reply:
You wouldn't need the TestPackageFile for most of the tests. 
You would only pass strings, no real file are needed.

Reply:
That would result in a change in the interface. I could write a helper function in the rule and that would accomplish the same thing, but we wouldn't be testing the `Validate` method anymore.